### ISSUE TITLE
Add security group id to be able to filter log messages in a multitenant environment

### ISSUE THREAD

**Issue by softinn72** [OraOpenSource/Logger#228]
Hello,
we have built several multitenant apex applications and we are in the process of switching to logger as our logging framework, but we are wondering how we can log an unique customer identifier like security_group_id to be able to distinguish the logging information coming from the same application when we have enabled debugging for 2 (or more) different customers at the same time. 

Adding this information in the extra column is easy enough, but then querying it back from a clob becomes complex and slow. We could also try to derive the customer identifier from the session client identifier or session username, but in case of scheduler jobs started from Apex things start becoming interesting. 

Another option we have evaluated is storing our security_group_id in the scope column, but in that case we would need to change all of our existing code to replace the logger.log% calls with our own local version to include the appropriate context we need.

Would it be an idea to add a new field to the logger_logs table like security_group_id we can fill in using a specific application context specified in the logger preferences? 

Then it would become really easy to query logging information for a specific customer and application user, without having knowledge of the internals of the application to retrace the session to a specific customer.

If not, do you have any better suggestions?

Thanks,
Paolo Marzucco

**Comment by jeffreykemp** [comment]
+1 I think it would be a great idea to add a "security_group_id" column to logger.

Logger would need a single API call to set (or unset) security_group_id at the start and end of a session.

**Comment by martindsouza** [comment]
@softinn72 Interesting idea. A few questions/comments regarding the need for the column (asking to get a better understanding to see if we have other options):

- Why do you need to separate logs for each tenant? I.e. are you trying to debug an error? Or do you trace through their steps on a frequent basis for analysis?
- I agree using the extra column is not a good idea as it's leveraged for other items as well

Overall I think asking the question "why" is important. My guess is that you may be using it for debugging issues that occur in APEX. What I do is create a custom APEX error handling function that gives me the starting point and I can trace backwards from there. This solution is independant of any multi-tenancy. I can describe it better if this is the root-cause of the request.

**Comment by softinn72** [comment]
Hello @martindsouza ,
thanks for your feedback.

In fact it's not about a specific issue to be solved, but more to give our first line helpdesk a way to sort easily through the logging for a specific customer once it's enabled. This will help them identify the error message related to that customer and decide if it's something they can handle with the internal knowledge base or if they need to forward the issue to our tech helpdesk. 

You'll understand that in case there are 2 customers which need to enable debugging at the same time, it won't be easy for the first line helpdesk to distinguish which log information belongs to whom. Having a security_group_id in the logger logs table will enable us to build a view on top of it to show only the logging pertaining to the specific customer they are trying to help.

I hope the issue is clear,
Paolo

**Comment by martindsouza** [comment]
@softinn72 Do each of these customers have different usernames (i.e. `app_user`) or can they both share the same `app_user` (just differentiated by tenant ID). In either case I think you can do what you want easily. I'll wait for your answer to detail it out.

**Comment by softinn72** [comment]
@martindsouza ,
currently we do have several applications with different constraints on usernames for each of them, so in theory there could be username overlaps across different applications, but as I said earlier things get interesting when starting background jobs via the database scheduler (which we do use heavily). Could your solution keep that into account?

**Comment by martindsouza** [comment]
Overall you can set the `client_identifier`. [this](https://www.talkapex.com/2013/05/logger-200-enable-session-specific/) article covers it.

**Comment by jeffreykemp** [comment]
I believe client_identifier is unsuitable as it identifies a particular client, not a tenant. For example, in a multi-tenant APEX application, client_identifier tells me the username and APEX session ID; I can't overwrite this with the tenant identifier.

**Comment by martindsouza** [comment]
@jeffreykemp Yes and no. Again it goes back to why you'd need to know which tenant you're dealing with.

For APEX: If you're investigating an error you can always create a manual APEX error handling function that logs all the APEX Items. One of the APEX items should be the `tenant_id`. If you want to always have this mapping (APEX session ID to tenant_id) just have a `onlogin` process to store the `session_id` with the `tenant_id` so you can always, and easily, back trace it.

For background jobs, if they are tenant specific (most likely not but just in case) you could have a "startup procedure" which both sets the Client Identifier and log the current `tenant_id`. If you have a good practice on how you name your Client IDs you can easily back trace these.

I know this is a bit of extra work for both of you. I'm always very resistant to adding new parameters to Logger as it's one more thing to support / manage. I really think you can get around the `tenant_id` issue with a bit of up-front planning. 

Please let me know if my solutions above will work for you.

**Comment by softinn72** [comment]
@martindsouza 
thanks for your feedback and I am sorry for my late reply, I've been very busy lately... 

I believe the above suggestion would be good enough only for a one-off analysis of the logs for a specific incident and it's not very useful to build a systematic approach wrt logging in our multitenant applications (like having a view on top of the logger tables to show only the logging pertaining to the specific customer/tenant). 

We are still discussing the approach to take to fully implement logger in all of our business applications, but at this point I believe that we are left with only 2 options:
1) build a logging wrapper that would add a tenant_id to the beginning of the scope column (if available in the current session), so that we could easily build a view filtering the logging information by tenant;
2) modify the locally installed logger source and add the missing tenant_id or security_group_id column by ourselves;

Of course there are cons for both options and we would have liked to have a better supported solution to avoid diverging too much from the actual project, but we also need to move forward with a workable solution.

I do understand your resistance to modify the existing code by adding support for this feature, but I also feel this would be a very welcome enhancement for several logger users which are implementing it in a complex business environment. I hope you may reconsider your decision in the future in case you will get more requests for it.

Any additional comment or suggestion is welcome.

Thanks for the attention,
Paolo Marzucco
